// yet-another-react-lightbox/dist/plugins/Inline.js
import * as React from "react";
import { clsx, createModule, cssClass } from "../core/index.js";
export const InlineContainer = ({ inline: { className, ...rest } = {}, children }) => (React.createElement("div", { className: clsx(cssClass("root"), className), ...rest }, children));
export const InlineModule = createModule("inline", InlineContainer);
export const Inline = ({ augment, replace, remove }) => {
    augment(({ toolbar: { buttons, ...restToolbar }, open, close, controller: { focus, aria, touchAction, ...restController }, className, ...restProps }) => ({
        open: true,
        close: () => { },
        toolbar: {
            buttons: buttons.filter((button) => button !== "close"),
            ...restToolbar,
        },
        inline: { style: { width: "100%", height: "100%" }, className },
        controller: { focus: false, aria: true, touchAction: "pan-y", ...restController },
        className,
        ...restProps,
    }));
    remove("no-scroll");
    replace("portal", InlineModule);
};
export default Inline;
